src/ipnet.rs

```diff
@@ -0,0 +1,828 @@
+
+    /// Create an Ipv6Net that contains *exclusively* the given address.
+    pub fn host_net(addr: Ipv6Addr) -> Self {
+        Self {
+            addr,
+            width: IPV6_NET_WIDTH_MAX,
+        }
+    }
+
+    /// Return the base address used to create this subnet.
+    pub fn addr(&self) -> Ipv6Addr {
+        self.addr
+    }
+
+    /// Return the prefix width.
+    pub fn width(&self) -> u8 {
+        self.width
+    }
+
+    pub(crate) fn mask(&self) -> u128 {
+        u128::MAX << (IPV6_NET_WIDTH_MAX - self.width)
+    }
+
+    /// Return the netmask address derived from prefix length.
+    pub fn mask_addr(&self) -> Ipv6Addr {
+        Ipv6Addr::from(self.mask())
+    }
+
+    /// Return true iff the subnet contains only the base address i.e. the
+    /// size is exactly one address.
+    pub fn is_host_net(&self) -> bool {
+        self.width == IPV6_NET_WIDTH_MAX
+    }
+
+    /// Return the number of addresses contained within this subnet.
+    pub fn size(&self) -> u128 {
+        1u128 << (IPV6_NET_WIDTH_MAX - self.width)
+    }
+
+    /// Return `true` if this subnetwork is in the IPv6 Unique Local Address
+    /// range defined in RFC 4193, e.g., `fd00:/8`
+    pub fn is_unique_local(&self) -> bool {
+        // TODO: Delegate to `Ipv6Addr::is_unique_local()` when stabilized.
+        self.first_addr().octets()[0] == 0xfd
+    }
+
+    /// Return the first address within this subnet.
+    pub fn first_addr(&self) -> Ipv6Addr {
+        let addr: u128 = self.addr.into();
+        Ipv6Addr::from(addr & self.mask())
+    }
+
+    /// The broadcast address for this subnet which is also the last address.
+    pub fn last_addr(&self) -> Ipv6Addr {
+        let addr: u128 = self.addr.into();
+        Ipv6Addr::from(addr | !self.mask())
+    }
+
+    /// Return an interator over the addresses of this subnet.
+    pub fn iter(&self) -> impl Iterator<Item = Ipv6Addr> {
+        let mask = self.mask();
+        let addr: u128 = self.addr.into();
+        let first = addr & mask;
+        let last = addr | !mask;
+        Ipv6NetIter {
+            next: Some(first),
+            last,
+        }
+    }
+
+    /// Return `true` if the address is within the subnet.
+    pub fn contains(&self, other: Ipv6Addr) -> bool {
+        let mask = self.mask();
+        let addr: u128 = self.addr.into();
+        let other: u128 = other.into();
+
+        (addr & mask) == (other & mask)
+    }
+
+    /// Return the nth address within this subnet or none if `n` is larger than
+    /// the size of the subnet.
+    pub fn nth(&self, n: u128) -> Option<Ipv6Addr> {
+        (n < self.size()).then(|| {
+            let addr: u128 = self.addr.into();
+            let nth = (addr & self.mask()) + n;
+            nth.into()
+        })
+    }
+
+    /// Returns `true` iff this subnet is wholly contained within `other`.
+    pub fn is_subnet_of(&self, other: &Self) -> bool {
+        other.addr <= self.addr && other.last_addr() >= self.last_addr()
```

Comment:
Why does the first comparison not use `first_addr()`? I'm not sure that's correct, but it's what I expected to see.

Reply:
Yeah that's probably righter.